pkg/distributors/contracts/DistributionScheduler.sol

```diff
@@ -17,104 +17,119 @@ pragma experimental ABIEncoderV2;
         for (uint256 i; i < scheduleIds.length; i++) {
             bytes32 scheduleId = scheduleIds[i];
             ScheduledDistribution memory scheduledDistribution = _scheduledDistributions[scheduleId];
 
+            // Silently skip any non-pending distributions as two users may be triggering two sets of
+            // scheduleIds with some overlap. This ensures that all distributions are started properly.
             if (scheduledDistribution.status != DistributionStatus.PENDING) {
                 continue;
             }
 
-            require(block.timestamp >= scheduledDistribution.startTime, "Distribution start time is in the future");
+            // Check that scheduled distribution is ready to be started.
 
+            require(block.timestamp >= scheduledDistribution.startTime, "Distribution start time is in the future");
             _scheduledDistributions[scheduleId].status = DistributionStatus.STARTED;
 
-            scheduledDistribution.distributionToken.approve(address(_multiDistributor), scheduledDistribution.amount);
+            // Send tokens to MultiDistributor and start distribution.
+
+            IERC20 distributionToken = _multiDistributor
+                .getDistribution(scheduledDistribution.distributionId)
+                .distributionToken;
+
+            distributionToken.approve(address(_multiDistributor), scheduledDistribution.amount);
             _multiDistributor.fundDistribution(scheduledDistribution.distributionId, scheduledDistribution.amount);
 
-            emit DistributionStarted(
-                scheduledDistribution.distributionId,
-                scheduleId,
-                scheduledDistribution.startTime,
-                scheduledDistribution.amount
-            );
+            emit DistributionStarted(scheduledDistribution.distributionId, scheduleId);
         }
     }
+
+    function cancelDistribution(bytes32 scheduleId) external override {
+        ScheduledDistribution memory scheduledDistribution = _scheduledDistributions[scheduleId];
+
+        // Check that scheduled distribution is eligible for cancellation.
+
+        require(scheduledDistribution.status != DistributionStatus.UNINITIALIZED, "Distribution does not exist");
+        require(scheduledDistribution.status != DistributionStatus.STARTED, "Distribution has already started");
+        require(
+            scheduledDistribution.status != DistributionStatus.CANCELLED,
+            "Distribution has already been cancelled"
+        );
+
+        _scheduledDistributions[scheduleId].status = DistributionStatus.CANCELLED;
+
+        // Check that caller is distribution owner and refund tokens.
+
+        IMultiDistributor.Distribution memory distributionChannel = _multiDistributor.getDistribution(
+            scheduledDistribution.distributionId
+        );
+
+        require(distributionChannel.owner == msg.sender, "Only distribution owner can cancel");
```

Comment:
The end result is the same, but for the benefit of the reader I'd move this require at the top of the function.

Reply:
This results in us getting an access control error on non-existent distributions rather than informing the user that the distribution doesn't exist.